enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+> An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM can provide a significantly better user experience by creating a channel for the operator to communicate state that OLM cannot infer otherwise.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate a condition that OLM cannot infer on its own.
+- Allow operators to define specific conditions which will influence OLM behavior, such as the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate conditions to OLM that it cannot infer otherwise.
+
+#### Story 2
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+Operators will communicate with OLM by setting [conditions](https://dev.to/maelvls/what-the-heck-are-kubernetes-conditions-for-4je7) on an owned CustomResource. Conditions are used by many Kubernetes Resources to provide users with additional insight into the state of the resource. For those unfamiliar with conditions, they typically follow the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type |  A one-word description of the condition. |
+| Status | Status is the status of the condition. Can be True, False, Unknown. |
+| Reason | Unique, one-word, CamelCase reason for the condition's last transition. |
+| Message | Human-readable message indicating details about last transition. Typically used by cluster admins for debugging purposes. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+
+### OLM Supported Conditions
+
+Although an operator may set any condition on a CustomResource that it owns, there will be a subset of "OLM Supported Conditions" which will influence the behavior of OLM when present. The list of "OLM Supported Conditions" can be seen in the table below:
+
+| Condition Name | Effect |
+| -------------- | ------ |
+| NotInterruptible | When true, OLM will not upgrade the operator. The pod associated with the operator may still be killed by external forces.|
+| Important | OLM will highlight this condition to the Cluster Admin.|
```

Comment:
What about signaling readiness/health or an error at install/upgrade?

Reply:
Sounds useful, I can update this proposal to include the readiness condition.

Reply:
> or an error at install/upgrade

This is vague. If an operator fails to install how will it write to the status? A failure post startup?

Reply:
How do operator maintainers decide between reporting via `Important` vs. setting an alert?